lib/core/test/unit/Cardano/Wallet/Api/ServerSpec.hs

```diff
@@ -265,3 +317,116 @@ errorHandlingSpec = describe "liftHandler and toServerError" $ do
+        it "ours and foreign keys are disjoint" $ do
+            property prop_listStakeKeysDisjoint
+
+prop_listStakeKeysDisjoint
+    :: UTxO
+    -> Set RewardAccount
+    -> (Map RewardAccount (Maybe Coin))
+    -> Property
+prop_listStakeKeysDisjoint utxo ours' m =
+    let
+        -- Build a list to pass listStakeKeys'
+        ours = zipWith (\ix acc -> (acc, ix, noDelegation))
+            [(0::Natural) ..]
+            (Set.toList ours')
+
+        ApiStakeKeys{_ours, _foreign, _none } = runIdentity $
+            listStakeKeys' utxo accountOfAddress (const $ pure m) ours
+
+        ourKeys = keys _ours
+        foreignKeys = keys _foreign
+    in
+        classify (length ourKeys > 1 && length foreignKeys > 1) "non-trivial" $
+            keys _ours `disjoint` keys _foreign
+  where
+    labeledCheck notF f a b = counterexample
+        (show a <> " " <> notF <> " " <> show b)
+        (property $ f a b)
+
+    disjoint = labeledCheck "is not disjoint from" Set.disjoint
+
+    keys = Set.fromList . map (view #_key)
+
+prop_listStakeKeysBalance
+    :: UTxO
+    -> Set RewardAccount
+    -> Property
+prop_listStakeKeysBalance utxo ours' = checkCoverage $
+    let
+        -- Build a list to pass listStakeKeys'
+        --
+        -- Passing the same account twice in the list would break the
+        -- property, but protecting against this in the implementation of
+        -- listStakeKeys' might not be worth it.
+        ours = zipWith (\ix acc -> (acc, ix, noDelegation))
+            [(0::Natural) ..]
+            (Set.toList ours')
+        ApiStakeKeys{_ours, _foreign, _none } = runIdentity $
+            listStakeKeys' utxo accountOfAddress (const $ pure mempty) ours
+
+        ourKeys = keys _ours
+        foreignKeys = keys _foreign
+        isNonTrivial =
+            length ourKeys > 1
+            && length foreignKeys > 1
+            && totalStake [_none] > 0
+
+    in
+       cover 35 isNonTrivial "non-trivial" $
+            ((totalStake _ours) +
+                (totalStake _foreign) +
+                (totalStake [_none]))
+                ===
+            getQuantity (coinToQuantity (TokenBundle.getCoin (balance utxo)))
+  where
+    keys = Set.fromList . map (view #_key)
+    totalStake = sum . map (getQuantity . view (#_stake))
+
+-- | Just something to pass listStakeKeys
+noDelegation :: ApiWalletDelegation
+noDelegation = ApiWalletDelegation
+    (ApiWalletDelegationNext NotDelegating Nothing Nothing)
+    []
+
+accountOfAddress :: Address -> Maybe RewardAccount
+accountOfAddress (Address bytes) =
+    let
+        str = B8.unpack bytes
+        char = last str
+    in
+        -- Leveraging the fact that addresses from genUTxOIndexSmall are
+        -- "ADDR{0 .. 7}"
+        if char == '7'
+        then Nothing
+        else Just $ RewardAccount $ B8.pack [char]
+
+instance Arbitrary UTxO where
+    arbitrary = toUTxO <$> genUTxOIndexSmall
+    shrink = map toUTxO . shrinkUTxOIndexSmall . fromUTxO
```

Comment:
Hi @Anviking 

I'm guessing that there's quite a bit of unnecessary repeated re-indexing going on here:
- `shrinkUTxOIndexSmall` first unwraps the index, shrinks the entries, and then recreates the index.
- `shrink` (defined above) again first calls `fromUTxO` to recreate the index, then calls `shrinkUTxOIndexSmall` (which reindexes), only to throw away the index again in the final step.

But the properties defined in this PR don't seem to make use of UTxO indexing functionality.

Perhaps we could avoid the indexing steps entirely?

I'm guessing that one way to do this would be to extract out the UTxO-generation part of `genUTxOIndexSmall` to a new function `UTxO.Gen.genUTxOSmall`, which `genUTxOIndexSmall` could reuse. And of course, the same for shrinkers.

What do you think?

Reply:
2d087558c27a2004c65b1cef7426ec3dd2255785 

I was worried there might be a performance hit to the UTxOIndex generators, but from trying RoundRobinSpec in ghci very briefly, I think we might be fine.